File: day/3/javascript/main.js
const getInput = require("../../../util/getInput.js");

const ALL_ASCII = [
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 65)),
];

const priority = ALL_ASCII.reduce(
  (acc, char, i) => ({ ...acc, [char]: i + 1 }),
  {}
);

const intersect2 = (xs, ys) => xs.filter((x) => ys.some((y) => y === x));
const intersection = (xs, ys, ...rest) =>
  ys === undefined ? xs : intersection(intersect2(xs, ys), ...rest);

(async function main() {
  const input = await getInput();
  const lines = input.split("\r\n");

  // Part 1
  let part_1_sum = 0;

  for (const line of lines) {
    const middle = Math.floor(line.length / 2);
    const [r1, r2] = [line.slice(0, middle), line.slice(middle)];
    part_1_sum += priority[intersection(Array.from(r1), Array.from(r2)).pop()];
  }
  console.log("Part 1:", part_1_sum);

  // Part 2
  let part_2_sum = 0;
  for (let i = 0; i < lines.length; i += 3) {
    const group = lines.slice(i, i + 3).map((line) => Array.from(line));
    part_2_sum += priority[intersection(...group).pop()];
  }
  console.log("Part 2:", part_2_sum);
})();

File: day/3/c++/main.cpp
#include "../../../util/CPPutil.h"
#include <iostream>
#include <unordered_map>
#include <vector>

std::unordered_map<char, int> createPriorities() {
    std::unordered_map<char, int> priorities;
    for (char c = 'a'; c <= 'z'; ++c)
        priorities[c] = c - 'a' + 1;
    for (char c = 'A'; c <= 'Z'; ++c)
        priorities[c] = c - 'A' + 26 + 1;
    return priorities;
}

int main() {
    auto priority = createPriorities();
    auto lines = ADVENT::getInputLines();

    size_t part1Sum = 0;

    for (auto &line : lines) {
        const size_t middle = line.size() / 2;
        auto r1 = ADVENT::sliceStr(line, 0, middle);
        auto r2 = ADVENT::sliceStr(line, middle, line.size());
        part1Sum += priority[ADVENT::intersection<char>({r1, r2})];
    }

    std::cout << "Part 1: " << part1Sum << std::endl;

    // Part 2
    size_t part2Sum = 0;

    for (int i = 0; i < lines.size(); i += 3) {
        std::vector<std::vector<char>> groups = {
            ADVENT::sliceStr(lines[i]),
            ADVENT::sliceStr(lines[i + 1]),
            ADVENT::sliceStr(lines[i + 2]),
        };
        part2Sum += priority[ADVENT::intersection<char>(groups)];
    }

    std::cout << "Part 2: " << part2Sum << std::endl;
}